File: 程式設計：資料結構答案檔/Python/105.py
class Term:
    def __init__(self, coefficient, x_exp, y_exp):
        self.coefficient = coefficient
        self.x_exp = x_exp
        self.y_exp = y_exp
        self.next = None

class Polynomial:
    def __init__(self):
        self.head = None

    def add_term(self, coefficient, x_exp, y_exp):
        if coefficient == 0:
            return
        new_term = Term(coefficient, x_exp, y_exp)
        if not self.head or (self.head.x_exp < x_exp or (self.head.x_exp == x_exp and self.head.y_exp < y_exp)):
            new_term.next = self.head
            self.head = new_term
            return
        prev = None
        current = self.head
        while current and (current.x_exp > x_exp or (current.x_exp == x_exp and current.y_exp > y_exp)):
            prev = current
            current = current.next
        if current and current.x_exp == x_exp and current.y_exp == y_exp:
            current.coefficient += coefficient
            if current.coefficient == 0:
                if prev:
                    prev.next = current.next
                else:
                    self.head = current.next
        else:
            new_term.next = current
            if prev:
                prev.next = new_term
            else:
                self.head = new_term

    def __str__(self):
        terms = []
        current = self.head
        while current:
            coeff_str = f"{current.coefficient}" if current.coefficient < 0 else f"+{current.coefficient}"
            terms.append(f"{coeff_str}x^{current.x_exp}y^{current.y_exp}")
            current = current.next
        return "".join(terms).lstrip('+')

    def add_polynomial(self, other):
        result = Polynomial()
        p = self.head
        while p:
            result.add_term(p.coefficient, p.x_exp, p.y_exp)
            p = p.next
        p = other.head
        while p:
            result.add_term(p.coefficient, p.x_exp, p.y_exp)
            p = p.next
        return result

input_polynomial_data = []
n = int(input())
for _ in range(n):
    input_polynomial_data.append(input())

polynomials = []
for poly_input in input_polynomial_data:
    poly = Polynomial()
    terms = poly_input.split()
    for i in range(0, len(terms), 3):
        coefficient, x_exp, y_exp = int(terms[i]), int(terms[i + 1]), int(terms[i + 2])
        poly.add_term(coefficient, x_exp, y_exp)
    polynomials.append(poly)

for i, poly_data in enumerate(input_polynomial_data):
    print(polynomials[i])

sum_poly = polynomials[0]
for poly in polynomials[1:]:
    sum_poly = sum_poly.add_polynomial(poly)

print(f"Sum:{sum_poly}")

File: 程式設計：資料結構答案檔/C#/105.cs
using System;
using System.Text;

class Term
{
    public int Coefficient { get; set; }
    public int XExp { get; set; }
    public int YExp { get; set; }
    public Term Next { get; set; }

    public Term(int coefficient, int xExp, int yExp)
    {
        Coefficient = coefficient;
        XExp = xExp;
        YExp = yExp;
        Next = null;
    }
}

class Polynomial
{
    public Term Head { get; private set; }

    public void AddTerm(int coefficient, int xExp, int yExp)
    {
        if (coefficient == 0) return;

        Term newTerm = new Term(coefficient, xExp, yExp);
        if (Head == null || Head.XExp < xExp || (Head.XExp == xExp && Head.YExp < yExp))
        {
            newTerm.Next = Head;
            Head = newTerm;
            return;
        }

        Term prev = null;
        Term current = Head;
        while (current != null && (current.XExp > xExp || (current.XExp == xExp && current.YExp > yExp)))
        {
            prev = current;
            current = current.Next;
        }

        if (current != null && current.XExp == xExp && current.YExp == yExp)
        {
            current.Coefficient += coefficient;
            if (current.Coefficient == 0)
            {
                if (prev != null)
                {
                    prev.Next = current.Next;
                }
                else
                {
                    Head = current.Next;
                }
            }
        }
        else
        {
            newTerm.Next = current;
            if (prev != null)
            {
                prev.Next = newTerm;
            }
            else
            {
                Head = newTerm;
            }
        }
    }

    public override string ToString()
    {
        StringBuilder result = new StringBuilder();
        Term current = Head;
        while (current != null)
        {
            string coeffStr = current.Coefficient < 0 ? $"{current.Coefficient}" : $"+{current.Coefficient}";
            result.Append($"{coeffStr}x^{current.XExp}y^{current.YExp}");
            current = current.Next;
        }
        return result.Length > 0 ? result.ToString().TrimStart('+') : "0";
    }

    public Polynomial AddPolynomial(Polynomial other)
    {
        Polynomial result = new Polynomial();
        Term p = this.Head;
        while (p != null)
        {
            result.AddTerm(p.Coefficient, p.XExp, p.YExp);
            p = p.Next;
        }
        p = other.Head;
        while (p != null)
        {
            result.AddTerm(p.Coefficient, p.XExp, p.YExp);
            p = p.Next;
        }
        return result;
    }
}

class Program
{
    static void Main()
    {
        int n = int.Parse(Console.ReadLine());
        Polynomial[] polynomials = new Polynomial[n];
        for (int i = 0; i < n; i++)
        {
            polynomials[i] = new Polynomial();
            string[] terms = Console.ReadLine().Split();
            for (int j = 0; j < terms.Length; j += 3)
            {
                int coefficient = int.Parse(terms[j]);
                int xExp = int.Parse(terms[j + 1]);
                int yExp = int.Parse(terms[j + 2]);
                polynomials[i].AddTerm(coefficient, xExp, yExp);
            }
        }

        foreach (var poly in polynomials)
        {
            Console.WriteLine(poly);
        }

        Polynomial sumPoly = polynomials[0];
        for (int i = 1; i < n; i++)
        {
            sumPoly = sumPoly.AddPolynomial(polynomials[i]);
        }

        Console.WriteLine($"Sum:{sumPoly}");
    }
}
